privacyidea/lib/tokenclass.py

```diff
@@ -470,18 +489,29 @@ def update(self, param, reset_failcount=True):
             raise ParameterError('[ParameterError] You may either specify '
                                  'genkey or otpkey, but not both!', id=344)
 
-        if otpKey is not None:
-            self.token.set_otpkey(otpKey, reset_failcount=reset_failcount)
-        elif genkey == 1:
-            otpKey = self._genOtpKey_()
+        if otpKey is None and genkey == 1:
+            otpKey = self._genOtpKey_(key_size)
 
         # otpKey still None?? - raise the exception
         if otpKey is None and self.hKeyRequired is True:
             otpKey = getParam(param, "otpkey", required)
 
         if otpKey is not None:
+            if self.token.rollout_state == "clientwait":
+                # If we have otpkey and the token is in the enrollment-state
+                # generate the new key
+                server_component = self.token.get_otpkey().getKey()
+                client_component = otpKey
+                otpKey = self.generate_symmetric_key(server_component,
+                                                     client_component)
+                self.token.rollout_state = ""
+                self.token.active = True
             self.add_init_details('otpkey', otpKey)
-            self.set_otpkey(otpKey)
+            self.token.set_otpkey(otpKey, reset_failcount=reset_failcount)
+
+        if two_step_init:
+            # After the key is generated, we set "waiting for the client".
+            self.token.rollout_state = "clientwait"
```

Comment:
We could think about to which extent we want to protect against API misuse: e.g. what happens when the second request also contains ``2stepinit=1``? Do we get a token which is active, but still has ``rollout_state=clientwait`` set?